app/services/vote_service.rb

```diff
@@ -18,19 +17,19 @@ def vote!(notifier = DebateNotifier.new(PusherBroadcaster))
 
   attr_reader :auth_token, :answer
 
-  def delete_previous_votes
-    previous_votes.destroy_all
-  end
-
-  def create_vote!
-    @answer.votes.create! auth_token: auth_token
+  def create_or_update_vote!
+    if previous_vote
+      previous_vote.update! answer_id: answer.id
+    else
+      answer.votes.create! auth_token: auth_token
+    end
   end
 
   def debate
     @auth_token.debate
   end
 
-  def previous_votes
-    @previous_votes ||= debate.votes.where(auth_token: auth_token)
+  def previous_vote
+    @previous_vote ||= debate.votes.find_by(auth_token: auth_token)
```

Comment:
it is an interesting case where normal caching will not work and in case of previous_vote being nil db query will be executed 2 times. https://karolgalanciak.com/blog/2017/05/28/ruby-memoization-%7C%7C-equals-vs-defined-syntax/ you can check this blogpost for details and possible fix.